concrete/src/Area/Layout/Preset/Provider/ActiveThemeProvider.php

```diff
@@ -1,34 +1,98 @@
 <?php
 namespace Concrete\Core\Area\Layout\Preset\Provider;
 
-use Page;
+use Concrete\Core\Page\Theme\Theme;
+use Concrete\Core\Http\Request;
+use Concrete\Core\Site\Service as SiteService;
 
 class ActiveThemeProvider implements ProviderInterface
 {
-    protected $presets = array();
+    /**
+     * The name of this provider.
+     *
+     * @var string
+     */
+    protected $name;
 
-    public function __construct()
+    /**
+     * The available layout presets.
+     *
+     * @var \Concrete\Core\Area\Layout\Preset\PresetInterface[]
+     */
+    protected $presets;
+
+    /**
+     * The theme that provides the presets.
+     *
+     * @var \Concrete\Core\Page\Theme\Theme|null
+     */
+    protected $theme;
+
+    public function __construct(Request $request, SiteService $siteService)
     {
-        $c = Page::getCurrentPage();
-        if (is_object($c)) {
+        $theme = null;
+        $c = $request->getCurrentPage();
+        if ($c && !$c->isError()) {
             $theme = $c->getCollectionThemeObject();
-            if (is_object($theme)) {
-                if ($theme instanceof ThemeProviderInterface) {
-                    $provider = new ThemeProvider($theme);
-                    $this->presets = $provider->getPresets();
-                    $this->name = $provider->getName();
-                }
+        }
+        if ($theme === null) {
+            $site = $siteService->getActiveSiteForEditing();
+            if ($site !== null) {
+                $themeID = $site->getThemeID();
+                $theme = $themeID ? Theme::getByID($themeID) : null;
             }
         }
+        if ($theme instanceof Theme) {
+            $this->theme = $theme;
+            $this->name = $theme->getThemeDisplayName();
+            if ($theme instanceof ThemeProviderInterface) {
+                $provider = new ThemeProvider($theme);
+                $this->presets = $provider->getPresets();
+            } else {
+                $this->presets = [];
+            }
+        } else {
```

Comment:
Can you explain a little bit about what this is solving?

Reply:
It was an attempt to solve the import failure, but it wasn't enough (see the changes in the CoreAreaLayout controller).
BTW I thought it was nice to have this change in this provider: it may help in case the current page has problems (for example: page not found).

Reply:
Sure, that makes sense to me. Sounds good!